### PR TITLE
2.0 proposal work-in-progress

### PR DESCRIPTION
**Pull request by kazooie**
I've tried to document most of the features we wanted for 2.0 here. Let me know if I missed anything or if you have other ideas/changes before I begin implementing.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -76,98 +127,91 @@ The standard state format for an input in redux looks like this:
...
 
 With this set up, you are able make changes to inputs and have them declaratively
  validated and state synchronized with your store. Because this could send actions
  to your store every keystroke, you probably want to use something like a
  [BlurInput](http://khan.github.io/react-components/#blur-input).
 
-By keeping these two redux-inputs functions flexible, it is easy to have multiple
- forms by making multiple inputConfigs. You can also make dynamic forms by
- dynamically creating input configs, replacing reducers, and creating dynamic components.
+## Action Creators/Thunks
```
Comment (kazooie) [inline README.md]: I'm considering we start passing these down as props to connectWithInput components in some way, so they are pre-bound to inputConfig and dispatch. The most common use case for using any of these right now is 
```
import { validateInputs } from 'redux-inputs'
...
const { dispatch } = this.props;
dispatch(validateInputs(inputConfig)).then(() => /* submit form */)
```

It could instead be something like
```
const { validateInputs } = this.props;
validateInputs().then(() => /* submit form */)
```

File: README.md [hunk 2]
```diff
@@ -45,25 +44,88 @@ Create an input configuration object and add a redux-inputs reducer to your redu
         inputs: createInputsReducer(inputConfig)
     });
 
-Note that `inputs` is the default mount point for redux-inputs state. You can
- configure this in your `inputConfig` object.
+### Step 4 - Connect
```
Comment (stonebk) [inline README.md]: I think this step's example needs some work, it's not very clear how it all works. So does `connectWithInputs` return a function similar to `connect`? Does a `mapStateToProps` and `mapDispatchToProps` fit in somewhere?

File: README.md [hunk 3]
```diff
@@ -45,25 +44,88 @@ Create an input configuration object and add a redux-inputs reducer to your redu
...
+other redux state specified:
 
-In this example your state would look like this:
+#### Props
 
-    {
-        loginModal: {
-            userInputs: { ... }
-        }
+- `inputProps` *{Object}* contains props for each input in the inputConfig
```
Comment (stonebk) [inline README.md]: What is this?

File: README.md [hunk 4]
```diff
@@ -45,25 +44,88 @@ Create an input configuration object and add a redux-inputs reducer to your redu
...
+
+- `inputActions` *{Object}*
```
Comment (stonebk) [inline README.md]: Rather than always passing these down as props, should we do something similar to `mapStateToProps` from redux? `connectWithInputs` would take `inputConfig` as well as a `mapInputActionsToProps` function? This would keep props clean on the component -- you'd only map those you needed.

File: README.md [hunk 5]
```diff
@@ -45,25 +44,88 @@ Create an input configuration object and add a redux-inputs reducer to your redu
...
+    - `inputActions.setInputs(newState, meta)`
+    - `inputActions.updateAndValidate(changes, meta)`
+    - `inputActions.validateInputs(inputKeys, meta)`
+    - `inputActions.resetInputs(inputKeys, meta)`
+    - `inputActions.initializeInputs(changes, meta)`
+    
+- `form` *{Object}*
```
Comment (stonebk) [inline README.md]: Same idea here, have a `mapFormStateToProps` function?

File: README.md [hunk 6]
```diff
@@ -45,25 +44,88 @@ Create an input configuration object and add a redux-inputs reducer to your redu
...
+    - `form.values` *{Object}* key-value pairs of the values of all inputs
+    - `form.validating` *{Boolean}* true if any inputs are currently async validating
+    - `form.pristine` *{Boolean}* true if all inputs are untouched, reset, or initialized.
+    
+
+### Step 5 - Wrapper
+
+import { InputsWrapper } from 'redux-inputs';
```
Comment (stonebk) [inline README.md]: Use code formatting here.

File: README.md [hunk 7]
```diff
@@ -45,25 +44,88 @@ Create an input configuration object and add a redux-inputs reducer to your redu
...
+    - `inputActions.resetInputs(inputKeys, meta)`
+    - `inputActions.initializeInputs(changes, meta)`
+    
+- `form` *{Object}*
+    - `form.values` *{Object}* key-value pairs of the values of all inputs
+    - `form.validating` *{Boolean}* true if any inputs are currently async validating
+    - `form.pristine` *{Boolean}* true if all inputs are untouched, reset, or initialized.
+    
+
+### Step 5 - Wrapper
```
Comment (stonebk) [inline README.md]: This needs a better example of how it is used.

File: README.md [hunk 8]
```diff
@@ -45,25 +44,88 @@ Create an input configuration object and add a redux-inputs reducer to your redu
...
+
+#### Props added to composed component
+
+- parser *(function)* transforms the new value before validating and updating
+- formatter *(function)* transforms the stored value before passing to the component
+- resolve *(function)* callback after successful change - same as resolve of updateAndValidate
+- reject *(function)* callback after failed validation - same as reject of updateAndValidate
+
+
+    import { ReduxInputsWrapper } from 'redux-inputs';
```
Comment (stonebk) [inline README.md]: This is not getting code formatting

File: README.md (hunk 1 above)
Comment (stonebk) [inline README.md]: Can you add a summary to this section to re-iterate what these are and where they fit in?

File: INPUTCONFIG.md [hunk 9]
```diff
@@ -2,41 +2,37 @@
...
-2. `inputsState` *(Object)* The current state of your inputs
+2. `inputs` *(Object)* The current state of your inputs
 3. `state` *(Object)* The state of the entire redux store
+4. `dispatch` *(Function)* dispatch function from the redux store, for dispatching side-effect actions
 
-#### Returns
-*(Boolean)*|*(Object)*: If boolean, true for valid, false for invalid. If object, represents the new state of the input.
-This can be used to add error information to the input state, such as setting errorText.
-
+#### Must return one of the following:
```
Comment (kazooie) [inline INPUTCONFIG.md]: Right now it would be a no-op with an error logged. Maybe it should assume false for nil values.

File: INPUTCONFIG.md [hunk 10]
```diff
@@ -50,50 +46,63 @@ This can be used to add error information to the input state, such as setting er
...
+ 
+#### Passed arguments
+1. `inputState` *(Object)* The new state of the input being changed. Contains properties such as `value`, `error`, `validating`.
+2. `inputs` *(Object)* The new state of your inputs
+3. `state` *(Object)* The new state of the entire redux store
+4. `dispatch` *(Function)* dispatch function from the redux store, for dispatching side-effect actions
+
+### `props` *(Object)* Props to pass down to every component associated with this input 
+
 ## Form Config
```
Comment (kazooie) [inline INPUTCONFIG.md]: The reducer needs to know which actions to listen to, and reducers don't know where they are mounted in the overall state.

File: README.md (hunk 3 above)
Comment (kazooie) [inline README.md]: The same object that `getInputProps` makes, bound to your inputConfig and dispatch.

File: README.md (hunk 4 above)
Comment (kazooie) [inline README.md]: All of the react-redux connect params are still optional in the connect function. 

The whole point of using this connect function over react-redux's connect is to pass down these additional props. There are only 3 of them, and the names for them can be configured (needs documentation) if there are name-collision issues. 

redux-form did the manual action mapping thing in some of the much earlier versions, but it was a lot more configuration for very little benefit, and they've just passed down everything for at least the last 3 major versions.

File: INPUTCONFIG.md [hunk 11]
```diff
@@ -50,50 +46,63 @@ This can be used to add error information to the input state, such as setting er
...
 
-You may set a `_form` property at the root of `inputConfig` to change some form-wide settings
```
Comment (kazooie) [inline INPUTCONFIG.md]: Change this to _reduxMountPoint only

File: README.md [hunk 12]
```diff
@@ -45,25 +44,86 @@ Create an input configuration object and add a redux-inputs reducer to your redu
...
+    
+    YourForm = connectWithMyForm(mapStateToProps, mapDispatchToProps, mergeProps, options)(YourForm);
+    
+Create a connect function for your form, which takes the `inputConfig` and creates a function that has the same 
+interface as [react-redux](https://github.com/reactjs/react-redux)'s `connect`. This will pass down these additional props to your component, in addition to any 
+other redux state specified:
+
+#### Props passed down to component
+
+- `inputProps` *{Object}* contains props for each input in the inputConfig
```
Comment (kazooie) [inline README.md]: Add explanation of how this ties in to Components

File: README.md (hunk 5 above)
Comment (jakepusateri) [inline README.md]: Consolidate the 3 props into 1 prop: 'form'. inputProps, setInputs, values, etc would all be on this prop. This leaves room for a single mapper: 'mapFormToProps' in the future

File: README.md [hunk 13]
```diff
@@ -45,25 +44,86 @@ Create an input configuration object and add a redux-inputs reducer to your redu
...
+- value: represents the current value as defined by the input's state
+- error *(boolean)*: True if the current value is invalid
+- errorText *(string)*
+- onChange *(function)*: Takes one parameter - the new value
+
+#### Props added to composed component
+
+- parser *(function)* transforms the new value before validating and updating
+- formatter *(function)* transforms the stored value before passing to the component
+- resolve *(function)* callback after successful change - same as resolve of updateAndValidate
```
Comment (jakepusateri) [inline README.md]: Rename resolve/reject to something more descriptive e.g. onValidationSuccess/onValidationError and fix callback api to call them with the same (input, inputs, state, dispatch) as all the other functions

File: README.md [hunk 14]
```diff
@@ -3,197 +3,60 @@
...
-In your connected component, use the `getInputProps` function to get an object with
- the following shape:
-    {
-        value: The last valid state the input had.
-        error: The state of the input if invalid, undefined otherwise. Should be displayed as value if present.
-        dispatchChange: A function to call on input changes to update the store
-    }
-
-The following example uses React, though it is not strictly required.
 ```jsx
```
Comment (stonebk) [inline README.md]: Comments for the single page example would be helpful.

File: docs/gettingStarted.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
+### Step 1 - Prerequisites
+
+Add [redux-thunk](https://github.com/gaearon/redux-thunk) middleware to your store,
+ following instructions at that repo.
+
+### Step 2 - [Config](./inputsConfig.md)
+
+Both the reducer and input props creator take a inputsConfig object that defines how inputs are initialized and validated.
```
Comment (stonebk) [inline docs/gettingStarted.md]: > Both the reducer and input props creator take...

This implies the use already knows about these, but this this the beginning of the get started. I would reword this to something like:

"The inputsConfig object defines how inputs are initialized and validated. It will be used when creating the inputs reducer and when connecting our input components."

File: docs/gettingStarted.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+### Example
+
+    const inputsConfig = {
+        homePrice: {
+            defaultValue: 300000,
+            validator: (value) => value > 0 && value < 350000000
+        }
+    }
+
+Create an input configuration object and add a redux-inputs reducer to your redux store in your project's reducers file.
```
Comment (stonebk) [inline docs/gettingStarted.md]: > ...and add a redux-inputs reducer to your redux store...

This is not shown in the Step 2 example at all. This should probably be part of Step 3.

File: docs/createInputsReducer.md [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    };
+    
+    const reducers = {
+        inputs: createInputsReducer(inputsConfig)
+    };
+    
+    const reducer = combineReducers(reducers);
+    const store = createStore(reducer);
+    
+### Use a different redux mount point
```
Comment (stonebk) [inline docs/createInputsReducer.md]: Make it more clear that this is optional. In most cases the user shouldn't have to worry about this at all.

File: docs/createInputsReducer.md [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+By default, redux-inputs assumes the reducer is mounted at `inputs`. For mounting in a different location, or 
+when multiple forms are stored in the state, use this special property in your inputsConfig object:
+
+#### `_reduxMountPoint` *(String)* change
```
Comment (stonebk) [inline docs/createInputsReducer.md]: `change` ?

File: docs/createInputsReducer.md [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+    {
+        tab1: {
+            inputs: {}
+        }
+    }
+
+### Multiple forms
+
+Create an inputsConfig for each form, making sure that each have a unique `_reduxMountPoint`.
```
Comment (stonebk) [inline docs/createInputsReducer.md]: It would be nice to explain why this is necessary. For examples, something like: "To avoid collisions at the default mount point, specify a custom mount point when using multiple forms."

File: docs/createInputsReducer.md [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+    };
+    
+    const reducers = {
+        formOne: createInputsReducer(inputsConfigOne)
+        formTwo: createInputsReducer(inputsConfigTwo)
+    };
+    
+    const reducer = combineReducers(reducers);
+    
+If your forms are similar and you want to share an input config, you can assign the common parts to each inputs config.
```
Comment (stonebk) [inline docs/createInputsReducer.md]: What?

File: docs/connectWithInputs.md [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+            );
+        }
+    }
+    
+    const connectWithMyForm = connectWithInputs(inputsConfig);
+    
+    YourForm = connectWithMyForm(mapStateToProps, mapDispatchToProps, mergeProps, options)(YourForm);
+    
+This will pass down redux inputs props to `YourForm`. By default, they are all added to a prop called `reduxInputs`, but 
+can be mapped to props using the `mapReduxInputsToProps` parameter.
```
Comment (stonebk) [inline docs/connectWithInputs.md]: An example of this would be helpful.

File: docs/ReduxInputsWrapper.md [hunk 22]
```diff
@@ -0,0 +1,42 @@
+## ReduxInputsWrapper(InputComponent:Component, [options:Object])
+
+This higher-order component is used to create components that can be used with `inputProps` from [connectWithInputs](connectWithInputs.md).
+
+#### Arguments
+- `InputComponent` *(Component)*
+- `options` *(Object)*
+- - `onChangeTransform` *(Function)* [optional] Turn a browser change event into a value. (e.g. `e => e.target.value` for `input`)
```
Comment (stonebk) [inline docs/ReduxInputsWrapper.md]: This is not being properly indented
